### PR TITLE
[Camera] Add new APIs for CameraDeviceManager and ExtraPreview

### PR DESCRIPTION
**Pull request by hsgwon**
### Description of Change ###

- Change internal CameraDeviceManager APIs to public.
- Add ExtraPreview APIs

### API Changes ###

 - TCSACR-498

### CODE REVIEW

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 1]
```diff
@@ -1237,9 +1237,8 @@ public CameraTagOrientation OrientationTag
         /// Gets the information of extra preview stream.
         /// </summary>
         /// <param name="streamId">The stream id.</param>
-        /// <since_tizen> 9 </since_tizen>
+        /// <since_tizen> 10 </since_tizen>
         /// <exception cref="ObjectDisposedException">The camera already has been disposed.</exception>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1241]: missing `<returns>`

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 2]
```diff
@@ -1265,9 +1263,8 @@ public void SetExtraPreviewStreamInfo(ExtraPreviewStreamInfo info)
         /// Gets the bitrate of extra preview with given stream id.
         /// </summary>
         /// <param name="streamId">The stream id.</param>
-        /// <since_tizen> 9 </since_tizen>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1268]: missing `<returns>`

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 3]
```diff
@@ -1281,21 +1278,60 @@ public int GetExtraPreviewBitrate(int streamId)
...
         {
             Native.SetExtraPreviewBitrate(_camera.GetHandle(), streamId, bitrate).
                 ThrowIfFailed("Failed to set extra preview bitrate");
         }
+
+        /// <summary>
+        /// Gets the GOP(Group Of Pictures) interval of extra preview with given stream id.
+        /// </summary>
+        /// <param name="streamId">The stream id.</param>
+        /// <since_tizen> 10 </since_tizen>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1293]: missing `<returns>`

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 4]
```diff
@@ -1281,21 +1278,60 @@ public int GetExtraPreviewBitrate(int streamId)
...
+        {
+            Native.SetExtraPreviewBitrate(_camera.GetHandle(), streamId, gopInterval).
+                ThrowIfFailed("Failed to set extra preview gop interval");
+        }
+
+        /// <summary>
+        /// Gets the rotation of extra preview with given stream id.
+        /// </summary>
+        /// <param name="streamId">The stream id.</param>
+        /// <since_tizen> 10 </since_tizen>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1320]: missing `<returns>`

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 5]
```diff
@@ -1319,40 +1354,35 @@ public ExtraPreviewStreamInfo(int streamId, CameraPixelFormat format, Size size,
...
-        /// <since_tizen> 9 </since_tizen>
-        [EditorBrowsable(EditorBrowsableState.Never)]
-        public int StreamId { get; set;}
+        /// <since_tizen> 10 </since_tizen>
+        public int StreamId { get; }
 
         /// <summary>
         /// Gets the extra preview format.
         /// </summary>
         /// <value></value>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1363]: empty `<value>`, either fill it in or remove it, same for the next 2 properties

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 6]
```diff
@@ -61,74 +58,79 @@ public CameraDeviceManager()
         /// <summary>
         /// Gets the status whether camera device(usb, network) is connected or not.
         /// </summary>
-        [EditorBrowsable(EditorBrowsableState.Never)]
+        /// <exception cref="ObjectDisposedException">The CameraDeviceManager already has been disposed.</exception>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:61]: missing `<returns>`

File: src/Tizen.Multimedia.Camera/Camera/CameraCapabilities.cs [hunk 7]
```diff
@@ -110,6 +111,13 @@ private bool CheckRangeValid(GetRangeDelegate func)
         /// <since_tizen> 4 </since_tizen>
         public bool IsContinuousCaptureSupported { get; }
 
+        /// <summary>
+        /// Gets the extra preview feature's supported state.
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraCapabilities.cs:115]: it's correct, but would sound better: "Gets the extra preview feature availability." or "Gets the extra preview feature support state."

File: src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs [hunk 8]
```diff
@@ -245,10 +245,9 @@ public partial class Camera : IDisposable, IDisplayable<CameraError>
         private event EventHandler<ExtraPreviewEventArgs> _extraPreview;
         /// <summary>
         /// An event that occurs once per frame when previewing.
-        /// Preview callback is registered when an user adds a callback explicitly to avoid useless P/Invoke.
+        /// ExtraPreview callback is registered when an user adds a callback explicitly to avoid useless P/Invoke.
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs:248]: callback is a term from native API, for C# use event/event handler, unless you are talking about native layer (mentioning P/Invoke suggests that), but in that case consider moving those details into `<remarks>`

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 9]
```diff
@@ -1281,21 +1278,60 @@ public int GetExtraPreviewBitrate(int streamId)
...
+                ThrowIfFailed("Failed to set extra preview gop interval");
+        }
+
+        /// <summary>
+        /// Gets the rotation of extra preview with given stream id.
+        /// </summary>
+        /// <param name="streamId">The stream id.</param>
+        /// <since_tizen> 10 </since_tizen>
+        /// <exception cref="ObjectDisposedException">The camera already has been disposed.</exception>
+        public Rotation GetExtraPreviewRotaion(int streamId)
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1322]: 🛑 typo in method name

**Review by wiertel — DISMISSED** [review]
There is a typo in method name: GetExtraPreviewRotaion()
Please keep the Copyright year as 2021, the file has been originally added then. You can add 2022 if you want: "Copyright (c) 2021, 2022 ..." or "Copyright (c) 2021-2022 ..."
Few suggestions for a global change:
   stream id/stream Id -> stream ID
device id/device Id -> device ID
The fps -> The FPS
The camera already has been disposed. -> The camera has already been disposed. /
 The camera has been already disposed.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 9 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1322]: Fixed.

File: src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs:248]: This is useless information for user so I'll remove it.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs [hunk 10]
```diff
@@ -1281,21 +1278,60 @@ public int GetExtraPreviewBitrate(int streamId)
         /// </summary>
         /// <param name="streamId">The stream id.</param>
         /// <param name="bitrate">The bitrate fo extra preview.</param>
```
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1280]: Fixed.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 11]
```diff
@@ -275,48 +276,42 @@ internal CameraDeviceInformation(CameraDeviceType type, CameraDevice device, str
...
         /// Gets the camera device Id.
         /// </summary>
         /// <value>The camera device id.</value>
-        /// <since_tizen> 9 </since_tizen>
-        [EditorBrowsable(EditorBrowsableState.Never)]
+        /// <since_tizen> 10 </since_tizen>
         public string Id { get; }
 
         /// <summary>
         /// Gets the number of extra stream.
```
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:304]: Fixed.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 12]
```diff
@@ -255,8 +257,7 @@ public struct CameraDeviceInformation
         /// <param name="numberOfExtraStream">The number of extra stream</param>
```
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:257]: Fixed.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 13]
```diff
@@ -61,74 +58,79 @@ public CameraDeviceManager()
...
-
-            return GetDeviceInformation(deviceList);
-        }
-
-        internal static bool IsSupported
+        /// <returns>
+        /// if camera device exist, returns list of <see cref="CameraDeviceInformation"/>; otherwise returns Enumerable.Empty.
+        /// </returns>
+        /// <exception cref="ArgumentException">Invalid enumeration.</exception>
+        /// <exception cref="ArgumentNullException">name or id is null.</exception>
```
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:75]: Actually, it's not `SupportedDevices` argument but CameraDeviceInformation argument and it's from native fw.
So, I didn't use `paramref` tag.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 14]
```diff
@@ -61,74 +58,79 @@ public CameraDeviceManager()
...
+        /// <summary>
+        /// Gets the current camera device information.
+        /// Retrieves all the supported camera devices and returns its information.
+        /// </summary>
+        /// <returns>
+        /// if camera device exist, returns list of <see cref="CameraDeviceInformation"/>; otherwise returns Enumerable.Empty.
+        /// </returns>
+        /// <see also="WebRTCState"/>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("This API is only for backward compatibility. Plaese use SupportedDevices instead")]
```
[the remark this answers is not in the document]
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:125]: This method is used by vd division now. so we can't deprecate it now.
I'll remove `Obsolete` attribute and move the description to `<remarks>` tag.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 1 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1241]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 2 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1268]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 3 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1293]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 4 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1320]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs (hunk 5 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraSettings.cs:1363]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs (hunk 6 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:61]: Added.

File: src/Tizen.Multimedia.Camera/Camera/CameraCapabilities.cs (hunk 7 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraCapabilities.cs:115]: I changed it to `Gets the extra preview feature support state` and other similar descriptions are changed also.

File: src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs (hunk 8 above)
Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/Camera.Events.cs:248]: I removed it.

File: src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs [hunk 15]
```diff
@@ -117,12 +118,12 @@ public IEnumerable<CameraDeviceInformation> SupportedDevices
         /// Gets the current camera device information.
         /// Retrieves all the supported camera devices and returns its information.
         /// </summary>
+        /// <remarks>This method is only for backward compatibility. Plaese use SupportedDevices instead</remarks>
```
Comment (wiertel) [inline src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs:121]: typo

Comment (hsgwon) [reply src/Tizen.Multimedia.Camera/Camera/CameraDeviceManager.cs]: Fixed

**Review by aferin — APPROVED** [review]

**Pull request merged**